### ISSUE TITLE
Mon pod is stuck with a node-selector of a terminated node

### ISSUE THREAD

**Issue by pre** [rook/rook#6663]
Similar issues, which are all closed and say this issue is fixed:
* https://github.com/rook/rook/issues/2570
* https://github.com/rook/rook/pull/2631

How to "enforce" reconciliation for distributing the mon pods by the rook-ceph-operator?

**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**

SItuation: I'm testing different failure scenarios in a test cluster with a fresh install of rook/ceph:v1.4.7 and ceph/ceph:v15.2.5.

Steps:
* I recycled all the three mon nodes.
* All old nodes were drained and pods were scheduled to the new nodes.
* All three old nodes are terminated and removed from the cluster.
* Two mon pods (mon-a, mon-b) were scheduled on the new nodes.
* One mon pod (mon-c) is Pending because it is being scheduled with Node Selector: OLD_NODE (and OLD_NODE has terminated and is no longer available).

How to trigger reconciliation so that the third mon-c pod will be scheduled to a new node? Ie. how to tell to rook-ceph-operator that "Hey, the node selector of mon-c is stale, please update the node selector" ?

**Expected behavior:**

When nodes are cycled, all the mon nodes are scheduled to the new nodes. Mon nodes should not have a node-selector from a drained & terminated node.

**How to reproduce it (minimal and precise):**

* Install a fresh cluster.
* Recycle all the mon nodes. 

**File(s) to submit**:

* Cluster CR (custom resource), typically called `cluster.yaml`, if necessary
* Operator's logs, if necessary
* Crashing pod(s) logs, if necessary

 To get logs, use `kubectl -n <namespace> logs <pod name>`
When pasting logs, always surround them with backticks or use the `insert code` button from the Github UI.
Read [Github documentation if you need help](https://help.github.com/en/articles/creating-and-highlighting-code-blocks).

**Environment**:
* OS (e.g. from /etc/os-release): Centos 8
* Kernel (e.g. `uname -a`): 5.4.74-36.135.amzn2.x86_64
* Rook version (use `rook version` inside of a Rook Pod): rook/ceph:v1.4.7
* Storage backend version (e.g. for ceph do `ceph -v`): ceph/ceph:v15.2.5
* Kubernetes version (use `kubectl version`): 1.18
* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift): EKS
* Storage backend status (e.g. for Ceph use `ceph health` in the [Rook Ceph toolbox](https://rook.io/docs/rook/master/ceph-toolbox.html)):

HEALTH_ERR 1 filesystem is degraded; 1 filesystem is offline; 1 mds daemon damaged; 1 filesystem is online with fewer MDS than max_mds

The error `1 mds daemon damaged` is caused by `mon-c` being in a Pending state because it has a node-selector of the terminated node.

**Comment by pre** [comment]
After many minutes (at least 5 minutes) the `mon-c` was scheduled as `mon-d` to the new node. Cluster has now mon-a, mon-b and mon-d (there is no `mon-c` anymore).

However, cluster is still in an error state:
```
  cluster:
    id:     9aa4a06e-990b-47c9-a824-2228e29d16b3
    health: HEALTH_ERR
            1 filesystem is degraded
            1 filesystem is offline
            1 mds daemon damaged
            1 filesystem is online with fewer MDS than max_mds

  services:
    mon: 3 daemons, quorum a,b,d (age 26m)
    mgr: a(active, since 35m)
    mds: ceph:0/0 6 up:standby, 1 damaged
    osd: 3 osds: 3 up (since 37m), 3 in (since 37m)

  data:
    pools:   3 pools, 65 pgs
    objects: 3 objects, 22 B
    usage:   3.0 GiB used, 3.4 TiB / 3.4 TiB avail
    pgs:     65 active+clean
```

Message says that 6 OSDs are standby and 1 is damaged. 

However, I have only 6 OSD pods and 6 Nodes in total (3 osd + 3 mon) - there is no 7th Node so there shouldn't be a 7th MDS pdos either. 

The rook-ceph-operator should be able figure this out but it apparently can't.

**Comment by pre** [comment]
Same behaviour can be reproduces in a fresh cluster with Rook 1.5.0 

Steps: 
* Have a ceph cluster in a healthy state.
* Drain one Kubernetes Node , this will remove eg. `mon-a` pod. 
* The deployment of `mon-a` keeps node-selector of the old node and pod stays in Pending state. 

rook-ceph-operator says:

`2020-11-18 10:55:15.845507 W | op-mon: mon "a" not found in quorum, waiting for timeout (599 seconds left) before failover`

The new k8s node is already available and Pod could be scheduled there. However, the `mon-a` pod is `Pending` because it keeps the node-selector of the terminated node which has left the cluster. 

After waiting for 10 minutes (600 seconds), rook-ceph-operator performs a failover and a new `mon-d` is scheduled on the new node: 

`2020-11-18 11:05:51.011487 W | op-mon: mon "a" NOT found in quorum and timeout exceeded, mon will be failed over`

So - eventually things work out. However, 10 minutes is a long time to wait for nothing. 

Question:

How could rook-ceph-operator be "tipped" that the node-selector should be updated? This would either allow
* `mon-a` to be scheduled to the new k8s node immediately after the new node is available 
* `mon-d` be created without first having to wait for 10 minutes .

I'm looking for an automated solution which would remove the 10 minute wait.

**Comment by travisn** [comment]
If you expect the mons and osds to be portable between different nodes and you're backing rook with storage from the cloud provider that can be moved to another node, you'll want to follow the [pvc-based cluster](https://rook.github.io/docs/rook/v1.5/ceph-cluster-crd.html#pvc-based-cluster), which allows mons and osds to be portable to new nodes. In that case, node selectors would not be applied by rook. See also the [cluster-on-pvc.yaml example](https://github.com/rook/rook/blob/release-1.5/cluster/examples/kubernetes/ceph/cluster-on-pvc.yaml).

**Comment by pre** [comment]
>  pvc-based cluster

I'm using the `kernel` mounter due to performance. The pvc-based cluster could be a solution to a different problem.

What I'm trying to achieve is resiliency: I'm after a setup where a failed k8s node could be automatically replaced and the CephCluster would migrate to the new node as quicky as possible. 

This does not seem to be the case at the moment. The rook-ceph-operator hangs both `mon` *and* `osd` pods with the old node-selector:
*  With `mon` pods the new pod is not scheduled to the new node until the 10min (default) timeout has passed. The old `mon` pod hangs in `Pending` state with the old node's node-selector until the 10min timeout has passed.
* With `osd` pods the new `osd` pod is *not started at all* due to the old `osd` pod being in `Pending` state due to having a stale node-selector. The old `osd` pod stays `Pending` until it is manually deleted.

Hence - if a k8s node is replaced, manual steps are required. However, with more resilient orchestrating from the rook-ceph-operator, CephCluster could survive automatically and/or faster. 

A coerce solution would be something like this:
* rook-ceph-operartor watches for Kubernetes Events of Node resources.
* when a Node resource gets a taint or label due to the Node being drained, 
* then rook-ceph-operator knows that the node-selector of the mon/osd pod needs to be updated and migration process needs to be kicked in. 
* When I have the new replacement Node available at the time of the Drain, 
** CephCluster would not be in a warning state at all (now it's in warning state for 10minutes for `mon` and until manual intervention for `osd`) 
** migration could happen without downtime as  `osd` could resync to the new node before the pod is evicted from the old node
** the `osd` pod would have a finalizer disallowing the evict on drain until `osd` has synced all of its data  (osd would not cause a warning state for CephCluster) 

What happens now:
* The migration process now happens automatically with `mon` process but not until the 10min timeout has passed. I don't think the timeout should be made shorter since the timeout serves a different purpose. As a result, CephCluster is in a warning state for 10 minutes.
* The migration process does *NOT* happen automatically with `osd` process. Instead, CephCluster will be in an unhealthy state until *manual* steps are performed: 
** Delete the previous `osd` (eg. `osd-a`) pod which is in `Pending` state because of the stale node-selector. 
** After old `psd` pod is deleted, rook-ceph-operator will schedule a new `osd` pod (eg. `osd-d`) and it will join the CephCluster. 
** Cluster is now in warning state as it has 4 osds: 3 functioning and 1 unavailable (`osd-a`). 
** Old `osd-a` needs to be manually removed: 

```
# Perhaps rebalance and wait for it to complete.
$ ceph osd crush reweight osd.<ID> 0.0

# Remove the OSD (`osd-a`) :
$ ceph osd out <ID>
$ service ceph stop osd.<ID>
$ ceph osd crush remove osd.<ID>
$ ceph auth del osd.<ID>
$ ceph osd rm <ID>
```

**Comment by pre** [comment]
Same goal (drain node) with different symptoms: https://rook-io.slack.com/archives/C46Q5UC05/p1606292757492500

**Comment by travisn** [comment]
@pre In Rook we made the decision to never automatically remove OSDs as described [here](https://rook.github.io/docs/rook/v1.5/ceph-osd-mgmt.html#remove-an-osd) since we don't want to risk ever accidentally removing data. 

While it sounds ideal to make rook really smart to detect a new node, detect the drained node, then migrate OSDs from one node to another, it doesn't follow the design for OSDs in Rook. Instead, the pattern is:
1. Rook automatically adds OSDs on new nodes when detected. (I believe there is still a timing issue with this and it only works sometimes, see also #6214).  
1. Ceph automatically marks OSDs `out` if down for some amount of time (5-10 minutes). As long as there is spare capacity, Ceph will automatically repair/backfill the data from the `out` OSD on other available OSDs in the cluster that meet the CRUSH rules. 
1. Manual intervention is needed to clean up the dead OSD(s) as described previously.

**Comment by pre** [comment]
> the pattern is:
> 
> 1. Rook automatically adds OSDs on new nodes when detected. (I believe there is still a timing issue with this and it only works sometimes, see also #6214).
> 2. Ceph automatically marks OSDs `out` if down for some amount of time (5-10 minutes). As long as there is spare capacity, Ceph will automatically repair/backfill the data from the `out` OSD on other available OSDs in the cluster that meet the CRUSH rules.

The current behaviour does not get this far. The old osd pod stays in `Pending` state because it has a stale node-selector. Hence, the new osd pod is not scheduled without first *restarting* the rook-ceph-opetator. Only after operator is restarted, the new osd pod is scheduled (i waited for 120 minutes so should not be a timeout issue). 

After the operator is (manually) restarted, the behaviour follows this described pattern. 

Same with mon pods: the old mon pod with a stale node-selector stays in Pending, but with mon, the operator will schedule a new mon pod after the 10min timeout has passed. 

I feel the rook-ceph-operator does not behave in a resilient manner. 


> 3. Manual intervention is needed to clean up the dead OSD(s) as described previously.

That’s ok, but the party stops already many steps before this step can reached.

**Comment by travisn** [comment]
> The current behaviour does not get this far. The old osd pod stays in `Pending` state because it has a stale node-selector. Hence, the new osd pod is not scheduled without first _restarting_ the rook-ceph-opetator. Only after operator is restarted, the new osd pod is scheduled (i waited for 120 minutes so should not be a timeout issue).

You see the OSD restarted after the operator is restarted? Or do you mean that a new OSD is created on the new node after the operator is restarted? This is the issue I was referring to in #6214 that a new node is not detected reliably.

> Same with mon pods: the old mon pod with a stale node-selector stays in Pending, but with mon, the operator will schedule a new mon pod after the 10min timeout has passed.
> 
> I feel the rook-ceph-operator does not behave in a resilient manner.

The mons are ok after 10 minutes, right? So you are just concerned about the OSD? If we fix #6214 it sounds like this would address the concern.

**Comment by telmich** [comment]
I believe this is still a problem with rook-operator 1.17.3 and it continues to degrade the cluster:

```
[15:45] blind:~% kubectl -n rook-ceph get pods | grep mon
rook-ceph-mon-ig-8665d97cb9-5g54w                     1/1     Running     0             23m
rook-ceph-mon-ii-58bcf555fd-6nhd2                     0/1     Pending     0             19s
rook-ceph-mon-io-8f48c4b5b-wgm9t                      1/1     Running     0             18m
rook-ceph-mon-it-6f7868655f-6gshn                     0/1     Pending     0             5m22s
rook-ceph-mon-iv-c79847b9d-4jlxd                      1/1     Running     0             11m
[15:45] blind:~% 

```

**Comment by telmich** [comment]
If someone in the future stumbles upon this:

* it seems the rook-operator reads out the configmap `rook-ceph-mon-endpoints`
* in that configmap you find monitor names + addresses
* just switching out the host names in there to actual nodes that exist makes it function, i.e. edits the deployments with the defined host names

I did not verify the address configuration in the same map, as I believe this is mostly relevant for ceph itself and the rook-operator probably updates the configmap at some point in the future.

**Comment by travisn** [comment]
@telmich Rook considers the configmap as the source of truth for the mon endpoint and node affinity. It's not generally supported to tweak that configmap. 

What are you trying to accomplish? You're trying to move a mon to a new node? That's not supported if the mon is running based on the `dataDirHostPath`. If you have nodes that are frequently being replaced, you should be using the [cluster-on-pvc](https://rook.io/docs/rook/latest-release/CRDs/Cluster/pvc-cluster/) solution, which will back the mon with a PV, which can potentially allow it to move between nodes.

**Comment by telmich** [comment]
Good morning @travisn !

The use case is very simple: we are using rook on real hardware and real hardware tends to fail over time, so servers are replaced over time. In a non-rook ceph cluster what we usually do is as soon as a node that hosts a monitor fails, we add a new one "somewhere else". The loss of the monitor database does not matter, as we are always running 3 to 5 monitors.

So the expectation for rook would be to behave similar:

If a node that is in the monitor configmap does not exist in the kube apiserver anymore, create a new monitor elsewhere and remove the dead monitor from the map

The hardware does not fail frequently, however running multiple clusters in multiple locations, one could say that this case happens maybe once or twice a month that some hardware is being replaced and a node that previously contained a monitor has to be taken out of service.

If this is not the default in rook, a setting on the line of `mon.addnewMonIfMonhostDoesNotExistAnymore: true` would be appreciated a lot.

I see where you are coming from, with the assumption that the monitor database should be "saved", but in the real world that does not work if a node or OS disk breaks.

**Comment by travisn** [comment]
@telmich You should already be seeing the mons failover when it goes down for some 10-20 minutes. See this topic on [mon failover](https://rook.io/docs/rook/latest-release/Storage-Configuration/Advanced/ceph-mon-health/#failing-over-a-monitor). Is that not working for you?

**Comment by telmich** [comment]
@travisn To be honest, we did not wait 20 minutes to verify. The main reason for that is that after upgrading the operator instantly updated / created 2 monitor deployments that went into pending state and our team decided that before the operator updates a third (3/5) monitor deployment, it is better to fix it.

So what I'd suggest is that if the operator updates the monitor deployment, that it checks whether the node written into the configmap does actually exist in the cluster and if not,  trigger the mon failover instantly, because it cannot be fixed by adding a nodeSelector on a non existing node.

**Comment by travisn** [comment]
@telmich Right, if the mon is stuck pending and its node does not exist, then the failover could happen immediately. Would you mind opening a new issue for that specific request? Thanks!
